Slack on a large iPad or using an external display is extremely large.

megangardner opened the pull request:
#### 8dbe225f029575e5d95866aa7bd1302507de12ee
<pre>
Slack on a large iPad or using an external display is extremely large.
<a href="https://bugs.webkit.org/show_bug.cgi?id=290492">https://bugs.webkit.org/show_bug.cgi?id=290492</a>
<a href="https://rdar.apple.com/138614711">rdar://138614711</a>

Reviewed by Tim Horton.

Slack has a set viewport that doesn&apos;t take the screen size of the
device into account. This means that it doesn&apos;t look great on anything
but a smaller iPad. Quirk slack to ignore this set viewport width, but
narrowly, so that if they change their code it will behave as expected.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldIgnoreViewportArgumentsToAvoidEnlargedView const):
(WebCore::handleSlackQuirks):
(WebCore::Quirks::determineRelevantQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:
* Source/WebCore/page/ViewportConfiguration.cpp:
(WebCore::ignoreViewportArgumentsToAvoidElargedView):
(WebCore::ViewportConfiguration::setViewportArguments):
(WebCore::ViewportConfiguration::description const):
* Source/WebCore/page/ViewportConfiguration.h:
(WebCore::ViewportConfiguration::setCanIgnoreViewportArgumentsToAvoidEnlargedView):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::setCanIgnoreViewportArgumentsToAvoidEnlargedViewIfNeeded):
(WebKit::WebPage::didCommitLoad):
(WebKit::WebPage::updateWebsitePolicies):

Canonical link: <a href="https://commits.webkit.org/292836@main">https://commits.webkit.org/292836@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eb5f62967fa7b0959898161725d46276a8a935df

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97177 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16800 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7011 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102258 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47702 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99222 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17093 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25250 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74026 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31230 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100180 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12911 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87884 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54373 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12666 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5759 "Passed tests") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47102 "Failed to compile WebKit") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82714 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5836 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104280 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24252 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17697 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83072 "Passed tests") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/24628 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84010 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82483 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20769 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4716 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17756 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24216 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29367 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24039 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27351 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25612 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7818,6 +7818,12 @@ static void setCanIgnoreViewportArgumentsToAvoidExcessiveZoomIfNeeded(ViewportCo
 
+static void setCanIgnoreViewportArgumentsToAvoidEnlargedViewIfNeeded(ViewportConfiguration& configuration, LocalFrame* frame, bool shouldIgnoreMetaViewport)
+{
+    if (auto* document = frame ? frame->document() : nullptr; document && document->quirks().shouldIgnoreViewportArgumentsToAvoidEnlargedView())
```
```suggestion
    if (RefPtr document = frame ? frame->document() : nullptr; document && document->quirks().shouldIgnoreViewportArgumentsToAvoidEnlargedView())
```

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8216,7 +8224,9 @@ void WebPage::updateWebsitePolicies(WebsitePoliciesData&& websitePolicies)
     setCanIgnoreViewportArgumentsToAvoidExcessiveZoomIfNeeded(m_viewportConfiguration, localMainFrame.get(), shouldIgnoreMetaViewport());
+    setCanIgnoreViewportArgumentsToAvoidEnlargedViewIfNeeded(m_viewportConfiguration, localMainFrame.get(), true);
 #endif
+
```
```suggestion
```

pxlcoder commented on Source/WebCore/page/ViewportConfiguration.cpp:
```diff
@@ -208,6 +218,9 @@ bool ViewportConfiguration::setViewportArguments(const ViewportArguments& viewpo
         adjustViewportArgumentsToAvoidExcessiveZooming(m_viewportArguments);
 
+    if (m_canIgnoreViewportArgumentsToAvoidEnlargedView)
+        ignoreViewportArgumentsToAvoidElargedView(m_viewportArguments);
```
```suggestion
        ignoreViewportArgumentsToAvoidEnlargedView(m_viewportArguments);
```

pxlcoder commented on Source/WebCore/page/ViewportConfiguration.cpp:
```diff
@@ -57,6 +57,16 @@ static inline void adjustViewportArgumentsToAvoidExcessiveZooming(ViewportArgume
     arguments.width = zoomedWidthFromArguments / arguments.zoom;
 }
 
+static inline void ignoreViewportArgumentsToAvoidElargedView(ViewportArguments& arguments)
```
```suggestion
static inline void ignoreViewportArgumentsToAvoidEnlargedView(ViewportArguments& arguments)
```

pxlcoder commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -2456,6 +2465,19 @@ static void handleSFUSDQuirks(QuirksData& quirksData, const URL& quirksURL, cons
+        return;
+
+    UNUSED_PARAM(quirksURL);
+    UNUSED_PARAM(documentURL);
```
You can just omit parameter names entirely, since they are not used in any code path.

hortont424 commented on Source/WebCore/page/ViewportConfiguration.cpp:
```diff
@@ -57,6 +57,16 @@ static inline void adjustViewportArgumentsToAvoidExcessiveZooming(ViewportArgume
+        return;
+
+    if (arguments.width == 1024)
+        arguments.width = 0;
```
I wonder if we should set minimumEffectiveDeviceWidth to 1024 here so that devices that are SMALLER than 1024 still use 1024 as the minimum? Should see what @whsieh thinks.

Like, our intent is really to blow out the explicit width on screens that are much bigger, but we don't mean to ever let them shrink *below* 1024.

whsieh replied:
Yeah, that makes sense to me. Or alternately, we could adjust the quirk so it only applies if the real device width (`m_viewLayoutSize`) exceeds 1024.

hortont424 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7914,6 +7920,8 @@ void WebPage::didCommitLoad(WebFrame* frame)
 
     setCanIgnoreViewportArgumentsToAvoidExcessiveZoomIfNeeded(m_viewportConfiguration, coreFrame.get(), shouldIgnoreMetaViewport());
+
+    setCanIgnoreViewportArgumentsToAvoidEnlargedViewIfNeeded(m_viewportConfiguration, coreFrame.get(), shouldIgnoreMetaViewport());
```
shouldn't this one just be false?

whsieh replied:
Hm... at navigation commit-time, I'd _expect_ the desktop website policy flag (which determines whether we opt in to `MetaViewportPolicy::Ignore`) to be set already, since that's determined at navigation decision time (or if the client sets the default webpage prefs through the configuration). I haven't tested this though..

megangardner replied:
If I set it to false then this doesn't work. I'm not super familiar with the shouldIgnoreMetaViewport() code, so if we think this is wrong I might need additional eye.

hortont424 replied:
I am confused, then. I would expect this bit to only be driven by the quirk, and nothing else.

hortont424 replied:
Wenson is right that I had the order wrong, but I'm confused why we care about `shouldIgnoreMetaViewport` at all for this quirk, I guess.

megangardner replied:
@hortont424 did you mean "true" instead of "false"? Cause that works. And it's what I'm doing in the other call as well.

hortont424 replied:
I misunderstood the flow a bit.

I *think* that you can just remove the boolean arg entirely, never consult 'shouldIgnoreMetaViewport', and always set it to true in setCanIgnoreViewportArgumentsToAvoidEnlargedViewIfNeeded in the case where the quirk is enabled.

(I still don't understand why the previous quirk was predicated on shouldIgnoreMetaViewport, maybe Wenson can enlighten us, but I don't think this one should be).

whsieh replied:
The intent was to limit this quirk to only when the desktop website is requested.

If the mobile site is requested on a large screen, then I'd thought it's probably best to return to the old (zoomed-in) behavior.

megangardner committed:
Slack on a large iPad or using an external display is extremely large.
https://bugs.webkit.org/show_bug.cgi?id=290492
rdar://138614711

Reviewed by Tim Horton.

Slack has a set viewport that doesn't take the screen size of the
device into account. This means that it doesn't look great on anything
but a smaller iPad. Quirk slack to ignore this set viewport width, but
narrowly, so that if they change their code it will behave as expected.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldIgnoreViewportArgumentsToAvoidEnlargedView const):
(WebCore::handleSlackQuirks):
(WebCore::Quirks::determineRelevantQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:
* Source/WebCore/page/ViewportConfiguration.cpp:
(WebCore::ignoreViewportArgumentsToAvoidElargedView):
(WebCore::ViewportConfiguration::setViewportArguments):
(WebCore::ViewportConfiguration::description const):
* Source/WebCore/page/ViewportConfiguration.h:
(WebCore::ViewportConfiguration::setCanIgnoreViewportArgumentsToAvoidEnlargedView):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::setCanIgnoreViewportArgumentsToAvoidEnlargedViewIfNeeded):
(WebKit::WebPage::didCommitLoad):
(WebKit::WebPage::updateWebsitePolicies):

Canonical link: https://commits.webkit.org/292836@main
